File: src/main/java/org/nest/commons/_visitor/ComparisonOperatorVisitor.java
package org.nest.commons._visitor;

import org.nest.commons._ast.ASTExpr;
import org.nest.spl.symboltable.typechecking.Either;
import org.nest.symboltable.symbols.TypeSymbol;
import org.nest.utils.AstUtils;

import static org.nest.spl.symboltable.typechecking.TypeChecker.isBoolean;
import static org.nest.symboltable.predefined.PredefinedTypes.getBooleanType;
import static org.nest.symboltable.predefined.PredefinedTypes.getIntegerType;
import static org.nest.symboltable.predefined.PredefinedTypes.getRealType;

/**
 * @author ptraeder
 */
public class ComparisonOperatorVisitor implements CommonsVisitor{

  @Override
  public void visit(ASTExpr expr) {
    final Either<TypeSymbol, String> lhsType = expr.getLeft().get().getType();
    final Either<TypeSymbol, String> rhsType = expr.getRight().get().getType();

    if (lhsType.isError()) {
      expr.setType(lhsType);
      return;
    }
    if (rhsType.isError()) {
      expr.setType(rhsType);
      return;
    }

    if (
        ((lhsType.getValue().equals(getRealType()) || lhsType.getValue().equals(getIntegerType())) &&
        (rhsType.getValue().equals(getRealType()) || rhsType.getValue().equals(getIntegerType())))
            ||
        (lhsType.getValue().getName().equals(rhsType.getValue().getName()))
            ||
        isBoolean(lhsType.getValue()) && isBoolean(rhsType.getValue())) {
      expr.setType(Either.value(getBooleanType()));
      return;
    }

    //Error message for any other operation
    final String errorMsg = "\""+AstUtils.toString(expr)+"\" - Only numeric types can be compared, SI types must match exactly.";
    expr.setType(Either.error(errorMsg));
  }

}

File: pynestml/nestml/ComparisonOperatorVisitor.py
#
# ComparisonOperatorVisitor.py
#
# This file is part of NEST.
#
# Copyright (C) 2004 The NEST Initiative
#
# NEST is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# NEST is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NEST.  If not, see <http://www.gnu.org/licenses/>.

"""
expression : left=expression comparisonOperator right=expression
"""
from pynestml.nestml.PredefinedTypes import PredefinedTypes
from pynestml.nestml.TypeChecker import TypeChecker
from pynestml.nestml.ErrorStrings import ErrorStrings
from pynestml.nestml.NESTMLVisitor import NESTMLVisitor
from pynestml.nestml.Either import Either
from pynestml.utils.Logger import Logger, LOGGING_LEVEL
from pynestml.utils.Messages import MessageCode


class ComparisonOperatorVisitor(NESTMLVisitor):
    def visitExpression(self, _expr=None):
        lhsTypeE = _expr.getLhs().getTypeEither()
        rhsTypeE = _expr.getRhs().getTypeEither()

        if lhsTypeE.isError():
            _expr.setTypeEither(lhsTypeE)
            return
        if rhsTypeE.isError():
            _expr.setTypeEither(rhsTypeE)
            return

        lhsType = lhsTypeE.getValue()
        rhsType = rhsTypeE.getValue()

        if ((TypeChecker.isReal(lhsType) or TypeChecker.isInteger(lhsType))
            and (TypeChecker.isReal(rhsType) or TypeChecker.isInteger(rhsType))) \
                or (lhsType.equals(rhsType) and TypeChecker.isNumeric(lhsType)) \
                or (TypeChecker.isBoolean(lhsType) and TypeChecker.isBoolean(rhsType)):
            _expr.setTypeEither(Either.value(PredefinedTypes.getBooleanType()))
            return

        # Error message for any other operation
        if (TypeChecker.isUnit(lhsType) and TypeChecker.isNumeric(rhsType)) \
                or (TypeChecker.isUnit(rhsType) and TypeChecker.isNumeric(lhsType)):
            # if the incompatibility exists between a unit and a numeric, the c++ will still be fine, just WARN
            errorMsg = ErrorStrings.messageComparison(self, _expr.getSourcePosition())
            _expr.setTypeEither(Either.value(PredefinedTypes.getBooleanType()))
            Logger.logMessage(_message=errorMsg, _code=MessageCode.SOFT_INCOMPATIBILITY,
                              _errorPosition=_expr.getSourcePosition(),
                              _logLevel=LOGGING_LEVEL.WARNING)
            return
        else:
            # hard incombatibility, cannot recover in c++, ERROR
            errorMsg = ErrorStrings.messageComparison(self, _expr.getSourcePosition())
            _expr.setTypeEither(Either.error(errorMsg))
            Logger.logMessage(_code=MessageCode.HARD_INCOMPATIBILITY,
                              _errorPosition=_expr.getSourcePosition(),
                              _message=errorMsg, _logLevel=LOGGING_LEVEL.ERROR)
            return
